### ISSUE TITLE
Roslyn fails to work when RSA+SHA1 is not available

### ISSUE THREAD

**Issue by omajid** [dotnet/roslyn#59880]
TLDR: .NET 6 + strong naming is now broken on RHEL 9

The [SHA-1](https://en.wikipedia.org/wiki/SHA-1#Attacks) cryptographic algorithm has had a number of attacks against it. [Some researchers have claimed to broken it completely](https://shattered.io/). [Microsoft itself has stopped using it in a number of scenarios](https://docs.microsoft.com/en-us/lifecycle/announcements/sha-1-signed-content-retired).

Now, a number of Linux distributions are starting to disable support for SHA1.

RHEL 9 has disabled it for a number of use cases: https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/9-beta/html/considerations_in_adopting_rhel_9/assembly_security_considerations-in-adopting-rhel-9#ref_considerations-security-crypto_changes-to-security and will probably disable them for more use-cases:

> In RHEL 9, SHA-1 usage is restricted in the DEFAULT system-wide cryptographic policy. With the exception of HMAC and DNSSec usage, SHA-1 is no longer allowed in TLS, DTLS, SSH, IKEv2 and Kerberos protocols. Individual applications not controlled by crypto policies are also moving away from using SHA-1 hashes in RHEL 9. 

Unfortunately, Rolsyn still uses SHA1 (and only SHA1) for strong name signing: https://github.com/dotnet/roslyn/blob/315c2e149ba7889b0937d872274c33fcbfe9af5f/src/Compilers/Core/Portable/PEWriter/SigningUtilities.cs#L22

That means .NET 6 is now broken on RHEL 9.

Building on RHEL 9 leads to a `Unhandled exception. Interop+Crypto+OpenSslCryptographicException: error:03000098:digital envelope routines::invalid digest` exception when trying to use SHA1.

I am looking into some workarounds to re-enable SHA1 this, but it does look like in certain security configuration (eg, FIPS 140) SHA1 is simply banned and the underlying crypto library on Linux (OpenSSL) will not allow SHA1 to be used. And no matter what's done, it does make out of the box default user-experience for .NET on RHEL 9 pretty bad if anything involves strong-naming.

I have a couple of questions:

1. Is strong-name signing use of SHA1 in a security context or not?

   https://docs.microsoft.com/en-us/dotnet/standard/assembly/create-use-strong-named says:

   > Do not rely on strong names for security. They provide a unique identity only.

   But isn't proving identity part of security (Authentication)?

2. If strong naming is not security relevant, would it be possible to add a fallback implementation if `System.Security.Cryptography.IncrementalHash` doesn't support SHA1?

4. Are there any plans to move away from SHA1?

   Even projects that were completely tied to SHA1, such as git, [are migrating off of SHA1](https://git-scm.com/docs/hash-function-transition/).  .NET/Roslyn is still tied to it, as far as I can tell. Are there any plans (short term or long term) to remove the use of SHA1 here?

**Version Used**: 

.NET 6

**Steps to Reproduce**:

1. On a CentOS Stream 9 container image (`quay.io/centos/centos:stream9`) with a recent enough version of OpenSSL 3.0 (`openssl-libs-3.0.1-8` or newer)
4. Using .NET 6, use any feature that requires strong-naming (eg, signing or verifying assemblies)
5. 

**Expected Behavior**:

Things work

**Actual Behavior**:

`Unhandled exception. Interop+Crypto+OpenSslCryptographicException: error:03000098:digital envelope routines::invalid digest`

**Comment by vcsjones** [comment]
> If strong naming is not security relevant, would it be possible to add a fallback implementation if System.Security.Cryptography.IncrementalHash doesn't support SHA1?

Just so to be clear, the problem isn't SHA1. `IncrementalHash` over `HashAlgorithmName.SHA1` and the `SHA1` classes continue to work in CentOS Stream 9. The issue lies in RSA over SHA1.

The `IncrementalHash` step is working fine, what is failing is the RSA `SignHash` when given SHA1:

https://github.com/dotnet/roslyn/blob/315c2e149ba7889b0937d872274c33fcbfe9af5f/src/Compilers/Core/Portable/PEWriter/SigningUtilities.cs#L27

This is why git is unaffected by this change. Git does not RSA sign anything, it just uses the hashing algorithm, so it won't be broken by this change in CentOS Stream 9.

**Comment by jaredpar** [comment]
> Is strong-name signing use of SHA1 in a security context or not?

No it is not for security. It is just to fulfill the ECMA-335 contract for providing identity to assemblies. 

It is not something we, or any other .NET Compiler, has a choice in. The ECMA-335 spec mandates SHA1 here and we must follow it. F# for example has the same behavior 

> Are there any plans to move away from SHA1?

That question is better asked on dotnet/runtime.

**Comment by jcouv** [comment]
As Jared pointed out, this is not narrowly a .NET Compiler issue, but rather an ECMA-335 issue. As far as roslyn goes, this is by-design at the moment.
I see that there's already an [issue/thread](https://github.com/dotnet/runtime/issues/65874) in the runtime repo which seems the better place to sort this out, so I'll go ahead and close this. Thanks

**Comment by tmds** [comment]
@jaredpar @jcouv @vcsjones 

I have a project with these attributes.

```xml
    <SignAssembly>true</SignAssembly>
    <AssemblyOriginatorKeyFile>..\..\sign.snk</AssemblyOriginatorKeyFile>
    <PublicSign>true</PublicSign>
```

This will continue to work because `PublicSign` doesn't use the `RSA+SHA1`, right?

What does a configuration look like that doesn't work? Is it `<PublicSign>false</PublicSign>`?
